pkg/controller/machineautoscaler/machineautoscaler_controller_test.go

```diff
@@ -142,3 +175,146 @@ func TestValidateReference(t *testing.T) {
 		})
 	}
 }
+
+func TestHandleTargetChange(t *testing.T) {
+	// A target which will not be fetchable via the API.
+	missingTarget := newMachineTarget("missing-target")
+
+	var testCases = []struct {
+		label     string
+		newTarget *MachineTarget
+		oldTarget *MachineTarget
+	}{
+		{
+			// MachineAutoscaler with no previous target should have the
+			// annotations added to the newly set target.
+			label:     "no previous target",
+			newTarget: newMachineTarget("no-previous-target"),
+			oldTarget: nil,
+		},
+		{
+			// MachineAutoscaler with missing previous target should have the
+			// annotations added to the newly set target.
+			label:     "bad previous target",
+			newTarget: newMachineTarget("no-previous-target"),
+			oldTarget: missingTarget,
+		},
+		{
+			// MachineAutoscaler with a previous target, and a new target which
+			// is missing, should still remove annotations on previous target.
+			label:     "bad new target",
+			newTarget: missingTarget,
+			oldTarget: newMachineTarget("previous-target"),
+		},
+		{
+			// MachineAutoscaler with both previous and new targets found.
+			label:     "good targets",
+			newTarget: newMachineTarget("new-target"),
+			oldTarget: newMachineTarget("previous-target"),
+		},
+	}
+
+	cfg := Config{
+		Namespace:           TestNamespace,
+		SupportedTargetGVKs: DefaultSupportedTargetGVKs(),
+	}
+
+	for _, tt := range testCases {
+		t.Run(tt.label, func(t *testing.T) {
+			ma := NewMachineAutoscaler()
+
+			maName := types.NamespacedName{
+				Namespace: ma.Namespace,
+				Name:      ma.Name,
+			}
+
+			objects := []runtime.Object{ma}
+
+			// Only add the old target if it's not meant to be missing.
+			if tt.oldTarget != nil && tt.oldTarget != missingTarget {
+				objects = append(objects, tt.oldTarget)
+			}
+
+			// Only add the new target if it's not meant to be missing.
+			if tt.newTarget != nil && tt.newTarget != missingTarget {
+				objects = append(objects, tt.newTarget)
+			}
+
+			r := newFakeReconciler(cfg, objects...)
+
+			// If there's a previous target, first reconcile the
+			// MachineAutoscaler with it set.
+			if tt.oldTarget != nil {
+				setTarget(ma, tt.oldTarget)
+
+				if err := r.client.Update(context.TODO(), ma); err != nil {
+					t.Errorf("Error updating MachineAutoscaler: %v", err)
+				}
+
+				r.Reconcile(reconcile.Request{NamespacedName: maName})
+
+				// Re-fetch the MachineAutoscaler.
+				if err := r.client.Get(context.TODO(), maName, ma); err != nil {
+					t.Errorf("Error fetching MachineAutoscaler: %v", err)
+				}
+			}
+
+			// Now set the new target and reconcile again.
+			setTarget(ma, tt.newTarget)
+
+			if err := r.client.Update(context.TODO(), ma); err != nil {
+				t.Errorf("Error updating MachineAutoscaler: %v", err)
```

Comment:
Can the remainder of the test continue if Update() fails?

Reply:
I suppose not. We could make this fatal. Though, in practice, with the fake client, this will never fail.

Reply:
Fixed. Errors on update or fetch are fatal now.